Replace old apperrors package and add error handling (#1408)
kyma-incubator/compass: docs/director/03-error-handling.md

# Error handling

Director API returns custom errors when there are problems with handling the request. 

## GraphQL errors

These are the GraphQL errors that may occur when calling the Director API:

| Error type             | Error code    |                            Description                                                                      |
|------------------------|---------------|-------------------------------------------------------------------------------------------------------------|
| `InternalError`        | `10`          | Specifies any error that cannot be handled in the Director. Such situation is unique and serious.           |

> | `NotFound`             | `20`          | Indicates that a given resource cannot be found and further processing is impossible.                       |

Reviewer: ```suggestion
| `NotFound`             | `20`          | Indicates that the given resource cannot be found and further processing is impossible.                       |
```

> | `NotUnique`            | `21`          | Indicates that a given resource is not unique.                                                              |

Reviewer: ```suggestion
| `NotUnique`            | `21`          | Indicates that the given resource is not unique.                                                              |
```

| `InvalidData`          | `22`          | Indicates that the input data is invalid with the reason described in the error message.                    | 
| `InsufficientScopes`   | `23`          | Indicates that the client doesn't have permission to execute the operation.                                 |
| `TenantRequired`       | `24`          | Indicates that a tenant is not found in the request, which is required to successfully handle the request.  |
| `TenantNotFound`       | `25`          | Indicates that the internal tenant is not found in the Director.                                            |

> | `Unauthorized`         | `26`          | Indicates that request cannot be authorized.                                                                |            

Reviewer: ```suggestion
| `Unauthorized`         | `26`          | Indicates that the request cannot be authorized.                                                                |            
```

> | `InvalidOperation`     | `27`          | Indicates that the operation is invalid, because of restrictions, eg.: cannot delete label definition, where the label is used.   |

Reviewer: ```suggestion
| `InvalidOperation`     | `27`          | Indicates that the operation is invalid because of certain restrictions, e.g.: you cannot delete a given label definition when the label is used.   |
```
`where` or `when`? Is `cannot delete label definition, where the label is used` an example of a situation or an example of an error message?

Author: Is an example of situation.

> The Graphql Error response have one additional field `extenstion` which contains `error_code` and `error` fields.

Reviewer: ```suggestion
The GraphQL Error response has one additional field `extension` which contains the `error_code` and `error` fields.
```

There was a typo in `extension`. I hope it only occurred here in the doc and not in code..?

Author: Yes, only in doc.

> See example of response with error:

Reviewer: ```suggestion
See an example of the response with an error:
```

```json
{
  "errors": [
    {

>       "message": "Error message  [key=value; key=value]",

Reviewer: ```suggestion
      "message": "Error message [key=value; key=value]",
```

Is the double space intentional here?

Author: No

| `NotFound`             | `20`          | Indicates that a given resource cannot be found and further processing is impossible.                       |
| `NotUnique`            | `21`          | Indicates that a given resource is not unique.                                                              |
| `InvalidData`          | `22`          | Indicates that the input data is invalid with the reason described in the error message.                    | 
| `InsufficientScopes`   | `23`          | Indicates that the client doesn't have permission to execute the operation.                                 |
| `TenantIsRequired`     | `24`          | Indicates that a tenant is not found in the request, which is required to successfully handle the request.  |
| `TenantNotFound`       | `25`          | Indicates that the internal tenant is not found in the Director.                                            |
| `Unauthorized`         | `26`          | Indicates that request cannot be authorized                                                                 |            

> | `InvalidOperation`     | `27`          | Indicates that the operation is not possible to be executed                                                 |

Reviewer 2: This description seems a bit too vague, pretty much every error can be an `InvalidOperation` error with this description :)
